### PR TITLE
Add more documentation for Transport UDFs

### PR DESCRIPTION
**Pull request by shardulm94**
This documentation was written prior to 4eecb75c8821230af102c98797c806dd98c77dd2 and hence may need to be refactored a bit. Also we may need to add more repositories in the gradle repositories block inside the user guides, the current set of repositories in the guide is insufficient and causes _dependency not found_ issues. I am trying to figure out the minimum set of repositories needed.

### CODE REVIEW

File: docs/writing-tests.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
+# Writing Tests for Transport UDFs
+
+Thank you for using Transport UDFs. See also [documentation index](/README.md#documentation)
+
+This guide takes you through writing a test for a Transport UDF using the unified testing framework.
```
Comment (wmoustafa) [inline docs/writing-tests.md]: writing tests

File: README.md [hunk 2]
```diff
@@ -13,8 +13,25 @@ or formats. Currently, Transport is capable of generating
 engine-artifacts for Spark, Hive, and Presto, and format-artifacts for
 Avro. Further details on Transport can be found in this [LinkedIn Engineering blog post](https://engineering.linkedin.com/blog/2018/11/using-translatable-portable-UDFs).
 
+## Documentation
+
+- Project information: [README.md](/README.md)
+  - Transport release notes: [docs/release-notes.md](/docs/release-notes.md)
+  - Contributing [TODO]
```
[the remark this answers is not in the document]
Comment (shardulm94) [reply README.md]: Yes

File: docs/authoring-transport-udfs.md [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+- Complex types (maps, arrays, structs) with generics
+    - [MapFromTwoArraysFunction](../transportable-udfs-examples/transportable-udfs-example-udfs/src/main/java/com/linkedin/transport/examples/MapFromTwoArraysFunction.java)
+    - [StructCreateByIndexFunction](../transportable-udfs-examples/transportable-udfs-example-udfs/src/main/java/com/linkedin/transport/examples/StructCreateByIndexFunction.java)
+- UDF overloading
+    - [`TopLevelStdUDF` Interface](transport-udfs-api.md#toplevelstdudf-interface)
+    - [NumericAddFunction](../transportable-udfs-examples/transportable-udfs-example-udfs/src/main/java/com/linkedin/transport/examples/NumericAddFunction.java)
+    - [NumericAddIntFunction](../transportable-udfs-examples/transportable-udfs-example-udfs/src/main/java/com/linkedin/transport/examples/NumericAddIntFunction.java)
+    - [NumericAddLongFunction](../transportable-udfs-examples/transportable-udfs-example-udfs/src/main/java/com/linkedin/transport/examples/NumericAddLongFunction.java)
+- Accessing side files in the UDF
```
[the remark this answers is not in the document]
Comment (shardulm94) [reply docs/authoring-transport-udfs.md]: Fixed

File: docs/authoring-transport-udfs.md [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+Now you should be able to see the UDF jar as well as platform-specific artifacts being built in the `build/libs` folder inside the module.
+For instructions on how to use these artifacts, see [Using Transport UDFs](using-transport-udfs.md).
+
+## More Complex UDF Examples
+
+- Complex types (maps, arrays, structs) with generics
+    - [MapFromTwoArraysFunction](../transportable-udfs-examples/transportable-udfs-example-udfs/src/main/java/com/linkedin/transport/examples/MapFromTwoArraysFunction.java)
+    - [StructCreateByIndexFunction](../transportable-udfs-examples/transportable-udfs-example-udfs/src/main/java/com/linkedin/transport/examples/StructCreateByIndexFunction.java)
+- UDF overloading
+    - [`TopLevelStdUDF` Interface](transport-udfs-api.md#toplevelstdudf-interface)
```
[the remark this answers is not in the document]
Comment (shardulm94) [reply docs/authoring-transport-udfs.md]: Done

File: docs/authoring-transport-udfs.md [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    - [NumericAddFunction](../transportable-udfs-examples/transportable-udfs-example-udfs/src/main/java/com/linkedin/transport/examples/NumericAddFunction.java)
```
[the remark this answers is not in the document]
Comment (shardulm94) [reply docs/authoring-transport-udfs.md]: Done

File: docs/authoring-transport-udfs.md [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+
+Thank you for using Transport UDFs. See also [documentation index](/README.md#documentation)
+
+This guide takes you through writing a simple UDF in a Gradle project.
+For better understanding, it is suggested that you read [Transport UDFs API](transport-udfs-api.md) either prior to or alongside this guide.
+
+## Adding Dependency
+
+Add the following to the `build.gradle` of the Gradle module in which you wish to develop your UDF.
+Get the latest version of "com.linkedin.transport.plugin" plugin from [Gradle Plugin Portal](https://plugins.gradle.org/plugin/com.linkedin.transport.plugin)
```
[the remark this answers is not in the document]
Comment (shardulm94) [reply docs/authoring-transport-udfs.md]: Resolved by using the old way of specifying plugin dependencies until we publish to gradle plugin portal.

File: docs/authoring-transport-udfs.md [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+
+```Gradle
+plugins {
+    id "java"
+    id "com.linkedin.transport.plugin" version "TODO"
```
[the remark this answers is not in the document]
Comment (shardulm94) [reply docs/authoring-transport-udfs.md]: Removed

File: docs/using-transport-udfs.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
+# Using Transport UDFs
+
+Thank you for using Transport UDFs. See also [documentation index](/README.md#documentation)
+
+This guide describes the procedure for using Transport UDFs in various supported platforms. If the UDF library uses the Transport Plugin, UDF artifacts will be automatically generated for each platform. These artifacts are distinguished by maven/ivy classifiers in addition to the original UDF artifact coordinates. Follow the below sections on how to identify the correct artifact/class for your platform and consequently how to use it in the platform.
```
[the remark this answers is not in the document]
Comment (shardulm94) [reply docs/using-transport-udfs.md]: I think this is relevant in this usage doc as the authors of the udf are different from the consumers of the udfs. Open to suggestions on how to better phrase it.

File: docs/using-transport-udfs.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    - [Hive](#hive)
+    - [Spark](#spark)
+    - [Presto](#presto)
+
+## Identifying platform-specific UDF artifacts
+
+### Platform-specific artifact file
+
+As mentioned above, the Transport Plugin will automatically generate artifacts for each platform.
+- Ivy/Gradle
```
[the remark this answers is not in the document]
Comment (shardulm94) [reply docs/using-transport-udfs.md]: Rephrased this section.  Open to suggestions on how to better phrase it.

File: docs/using-transport-udfs.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (wmoustafa) [reply docs/using-transport-udfs.md]: Sounds good. I am okay with the current structure.

**Pull request merged**